docs/abi.md

```diff
@@ -113,17 +113,37 @@ functions** as
 
 Blocks until data is available for reading from one of the specified channel
 handles, unless any of the channels is invalid, orphaned, or violates the
-[information flow control](/docs/concepts.md#labels). The channel handles are
-encoded in a buffer that holds N contiguous 9-byte chunks, each of which is made
-up of an 8-byte channel handle value (little-endian u64) followed by a single
-channel status byte. Invalid handles will have an `INVALID_CHANNEL`, `ORPHANED`,
-or `PERMISSION_DENIED` status, but `wait_on_channels` return value will only
-fail for internal errors or if the runtime is terminating.
+[information flow control](/docs/concepts.md#information-flow-control). The
+channel handles are encoded in a buffer that holds N contiguous 9-byte chunks,
+each of which is made up of an 8-byte channel handle value (little-endian u64)
+followed by a single channel status byte. Invalid handles will have an
+`INVALID_CHANNEL`, `ORPHANED`, or `PERMISSION_DENIED` status, but
+`wait_on_channels` return value will only fail for internal errors or if the
+runtime is terminating.
 
 - `param[0]: usize`: Address of handle status buffer
 - `param[1]: u32`: Count N of handles provided
 - `result[0]: u32`: Status of operation
 
+### `wait_on_channels_with_privilege`
```

Comment:
Just to check my understanding:
This is a (yet to be implemented) way to deliver error messages about channel sends safely, right? And it is safe because when Node A sends to Node B, the error messages will only be delivered if B downgrades the error back to A. Right?

Reply:
Not quite.  This is a (yet to be implemented) variant of our "do I have incoming messages?" entrypoint (cf. `poll(2)` or `select(2)`), which allows the putative reader to distinguish between "I can't tell if I've got incoming messages on this particular channel because I can't read from it (because labels)" and "I can tell if I've got incoming messages on this particular channel, but I'll need to use my privilege to read them".   (So for this, there's nothing flowing back to the sender of the message.)

Reply:
OK right makes sense. 

Separate question: on a channel write, what happens if the channel is full?

Reply:
There's no explicit limit to channel size, they just keep accepting messages (which we did explicitly to prevent back-pressure being a reverse signal). 

(I guess in practice the heap allocation could fail, but if we get into that situation I imagine that all sorts of things we be going wrong&hellip;)

Reply:
Keep accepting sounds like a good way to handle the expected/common/good case.

Trying to cause a heap allocation failure sounds like something an attacker might want to do. What if we try to prevent this by having a channel size limit that is small enough to guarantee heap allocation failure (which I suppose potentially depends on the machine and the WASM runtime configuration?) but still large enough to usually not be annoying. When this fixed channel size is filled, we can't throw an error to the sender, but we can drop the new message, write over old entries, write over new entries.